## Task 1: Data Loading, Cleaning & Exploratory Data Analysis

### 1.1 Load and inspect the dataset

In [1]:
!git clone https://github.com/Deepvejpara/Income_Classifier.git

Cloning into 'Income_Classifier'...
remote: Enumerating objects: 10, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 10 (delta 0), reused 10 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (10/10), 619.43 KiB | 5.25 MiB/s, done.


In [2]:
import os

In [5]:
# Changing dir to edit inside folder
os.chdir('/content/Income_Classifier')

In [6]:
import numpy as np
import pandas as pd

In [8]:
df = pd.read_csv('dataset/adult.csv')

In [9]:
df.shape

(48842, 15)

In [11]:
print(df.dtypes)

age                 int64
workclass          object
fnlwgt              int64
education          object
educational-num     int64
marital-status     object
occupation         object
relationship       object
race               object
gender             object
capital-gain        int64
capital-loss        int64
hours-per-week      int64
native-country     object
income             object
dtype: object


In [12]:
df.head()

,age,workclass,fnlwgt,education,educational-num,marital-status,occupation,relationship,race,gender,capital-gain,capital-loss,hours-per-week,native-country,income
0,25,Private,226802,11th,7,Never-married,Machine-op-inspct,Own-child,Black,Male,0,0,40,United-States,<=50K
1,38,Private,89814,HS-grad,9,Married-civ-spouse,Farming-fishing,Husband,White,Male,0,0,50,United-States,<=50K
2,28,Local-gov,336951,Assoc-acdm,12,Married-civ-spouse,Protective-serv,Husband,White,Male,0,0,40,United-States,>50K
3,44,Private,160323,Some-college,10,Married-civ-spouse,Machine-op-inspct,Husband,Black,Male,7688,0,40,United-States,>50K
4,18,?,103497,Some-college,10,Never-married,?,Own-child,White,Female,0,0,30,United-States,<=50K


In [25]:
# Collecting Categorical columns in cols
cols = df.select_dtypes('object').columns
cols

Index(['workclass', 'education', 'marital-status', 'occupation',
       'relationship', 'race', 'gender', 'native-country', 'income'],
      dtype='object')

In [26]:
# Removing extra Space
for col in cols:
  df[col] = df[col].str.strip()

### 1.2 Handle missing values

In [27]:
df.replace('?', np.nan, inplace=True)

In [31]:
print(df.isnull().sum())

age                   0
workclass          2799
fnlwgt                0
education             0
educational-num       0
marital-status        0
occupation         2809
relationship          0
race                  0
gender                0
capital-gain          0
capital-loss          0
hours-per-week        0
native-country      857
income                0
dtype: int64


In [32]:
df.dropna(inplace=True)

**Justification for Dropping Missing Values over Imputation:**

The missing values in this dataset (originally encoded as `?`) exist exclusively in categorical columns with many unique categories: `workclass` (~2,799 missing), `occupation` (~2,809 missing), and `native-country` (~857 missing).

Using mode imputation (replacing missing values with the most frequent category) would disproportionately inflate the most common category, introducing unnatural bias into the dataset. Because dropping the rows with any missing value only removes approximately 2,399 records—leaving us with a robust dataset of ~45,222 clean records—dropping them is a statistically safer and cleaner approach than imputation for this specific dataset.

### 1.3 Drop redundant columns